pytorch_lightning/callbacks/device_stats_monitor.py

```diff
@@ -43,33 +52,62 @@ class DeviceStatsMonitor(Callback):
         >>> trainer = Trainer(callbacks=[device_stats]) # doctest: +SKIP
     """
 
-    def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: Optional[str] = None) -> None:
-        if not trainer.loggers:
-            raise MisconfigurationException("Cannot use DeviceStatsMonitor callback with Trainer that has no logger.")
+    def __init__(self, cpu_stats: Optional[bool] = None) -> None:
+        self.cpu_stats = cpu_stats
 
-    def on_train_batch_start(
+    def setup(
         self,
         trainer: "pl.Trainer",
         pl_module: "pl.LightningModule",
-        batch: Any,
-        batch_idx: int,
-        unused: int = 0,
+        stage: Optional[str] = None,
     ) -> None:
         if not trainer.loggers:
             raise MisconfigurationException("Cannot use `DeviceStatsMonitor` callback with `Trainer(logger=False)`.")
+        # warn in setup to warn once
+        device = trainer.strategy.root_device
+        if self.cpu_stats is None and device.type == "cpu" and not _PSUTIL_AVAILABLE:
+            # TODO: raise an exception from v1.9
+            rank_zero_warn(
+                "`DeviceStatsMonitor` will not log CPU stats as `psutil` is not installed."
+                " To install `psutil`, run `pip install psutil`."
+                " It will raise an exception if `psutil` is not installed post v1.9.0."
+            )
 
+    def _get_and_log_device_stats(self, trainer: "pl.Trainer", key: str) -> None:
         if not trainer._logger_connector.should_update_logs:
             return
+        if not trainer.loggers:
+            raise MisconfigurationException("Cannot use `DeviceStatsMonitor` callback with `Trainer(logger=False)`.")
 
         device = trainer.strategy.root_device
+        if self.cpu_stats is None and device.type == "cpu" and not _PSUTIL_AVAILABLE:
```

Comment:
how about making `cpu_stats` a protected and marking it as False when the warning is raised so that you don't have to check the condition again here?

Reply:
Good suggestion. Updated!